components/org.wso2.carbon.identity.application.authenticator.fido2.endpoint/src/main/java/org/wso2/carbon/identity/application/authenticator/fido2/endpoint/impl/StartRegistrationApiServiceImpl.java

```diff
@@ -0,0 +1,78 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.carbon.identity.application.authenticator.fido2.endpoint.impl;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.authenticator.fido2.core.WebAuthnService;
+import org.wso2.carbon.identity.application.authenticator.fido2.dto.RegistrationRequest;
+import org.wso2.carbon.identity.application.authenticator.fido2.endpoint.StartRegistrationApiService;
+import org.wso2.carbon.identity.application.authenticator.fido2.endpoint.common.Constants;
+import org.wso2.carbon.identity.application.authenticator.fido2.exception.FIDO2AuthenticatorClientException;
+import org.wso2.carbon.identity.application.authenticator.fido2.util.Either;
+import org.wso2.carbon.identity.application.authenticator.fido2.util.FIDOUtil;
+import org.wso2.carbon.identity.core.util.IdentityCoreConstants;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import javax.ws.rs.core.Response;
+
+import static org.wso2.carbon.identity.application.authenticator.fido2.endpoint.common.Util.getErrorDTO;
+
+/**
+ * StartRegistrationApiServiceImpl class is used to trigger FIDO2 device registration.
+ */
+public class StartRegistrationApiServiceImpl extends StartRegistrationApiService {
+
+    private static final Log LOG = LogFactory.getLog(StartRegistrationApiServiceImpl.class);
+
+    @Override
+    public Response startRegistrationPost(String appId) {
+
+        if (StringUtils.isBlank(appId)) {
+            return Response.status(Response.Status.BAD_REQUEST).entity(getErrorDTO
+                    (Constants.ErrorMessages.ERROR_CODE_START_REGISTRATION_EMPTY_APP_ID)).build();
+        }
+
+        try {
+            if(appId.contains(Constants.EQUAL_OPERATOR)) {
+                appId = URLDecoder.decode(appId.split(Constants.EQUAL_OPERATOR)[1], IdentityCoreConstants.UTF_8);
+            }
+            WebAuthnService webAuthnService = new WebAuthnService();
+            Either<String, RegistrationRequest> result = webAuthnService.startRegistration(appId);
+            if (result.isRight()) {
+                return Response.ok().entity(FIDOUtil.writeJson(result.right().get())).build();
+            } else {
+                return Response.serverError().entity(getErrorDTO(Constants.ErrorMessages.ERROR_CODE_START_REGISTRATION,
+                        appId)).build();
+            }
+        } catch (JsonProcessingException e) {
+            LOG.error(e);
```

Comment:
Log exception along with a proper message explaining what operation get failed and include context information to separate out logs. Check this in all the places.

Reply:
Added.